File: Data Structures & Algorithms/maximum-subarray/submission-2.java
class Solution {
    public int dfs(int i , boolean flag , int[] nums){
        if(i==nums.length){
           if(flag) return 0;
           return -10000;
        }
        if(flag)
           return Math.max(dfs(i+1,true,nums)+nums[i],0);
        else
           return Math.max(nums[i]+dfs(i+1,true,nums),dfs(i+1,false,nums));
    }
    public int maxSubArray(int[] nums) {
        return dfs(0,false,nums);
    }
}

File: Data Structures & Algorithms/longest-turbulent-subarray/submission-2.py
class Solution:
    def maxTurbulenceSize(self, arr: List[int]) -> int:
        maxSize=1
        dp={}
        def dfs(i) :
            if arr[i]==arr[i+1]:
                return 1
            elif arr[i]>arr[i+1] :
                return 2 + fun(i+1,False)
            else:
                return 2 + fun(i+1,True)
        def fun(i,Greater):
            if (i,Greater) in dp:
                return dp[(i,Greater)]
            res=0
            if i==len(arr)-1 :
                return res
            elif Greater and  arr[i]>arr[i+1] :
                res= 1+fun(i+1,False)
            elif not Greater and arr[i]<arr[i+1] :
                res= 1+fun(i+1,True)
            dp[(i,Greater)]=res
            return res
        for i in range(len(arr)-1):
            maxSize=max(maxSize,dfs(i))
        return maxSize